src/engine/macros.rs

```diff
@@ -23,10 +23,14 @@ macro_rules! destroy_filesystems {
     ( $s:ident; $fs:expr ) => {
         let mut removed = Vec::new();
         for uuid in $fs.iter().map(|x| *x) {
+            match $s.filesystems.get_mut_by_uuid(&uuid) {
```

Comment:
But, it's also the case that none of the ```teardown()``` methods require a mutable self, which means that Filesystem's ```destroy()``` method does not require a mutable self either. I think those changes should be made, and then you can certainly use result of ```remove_by_uuid()``` method, regardless of any changes I can make.